backend/src/advert/advert.service.ts

```diff
@@ -0,0 +1,194 @@
+            .andWhere('advert.subjectId = :subjectId', {subjectId: advertFilter.subjectId})
+            .leftJoinAndSelect('advert.teacher', 'teacher')
+            .leftJoinAndSelect('advert.subject', 'subject');
+
+        if (advertFilter.city != null) {
+            queryBuilder = await queryBuilder.andWhere('');
+        }
+
+        if (advertFilter.place != null) {
+            queryBuilder.andWhere('advert.place = :place', {place: advertFilter.place});
+        }
+
+        if (advertFilter.level != null) {
+            queryBuilder.andWhere('advert.place = :level', {level: advertFilter.level});
+        }
+
+        if (advertFilter.priceTo != null) {
+            queryBuilder.andWhere('advert.price between :from and :to', {from: priceFrom, to: advertFilter.priceTo});
+        }
+
+        const entities = await queryBuilder
+            .skip((page - 1) * advertFilter.itemsPerPage)
+            .take(advertFilter.itemsPerPage)
+            .getMany();
+
+        const count = await queryBuilder.getCount();
+        const hasNext = (page + 1) * advertFilter.itemsPerPage < count;
+        const hasPrev = (page - 1) * advertFilter.itemsPerPage > 0;
+
+        return new PaginatedResult<AdvertDto>(entities.map(x =>
+                new AdvertDto(x.id,
+                    x.place,
+                    x.level,
+                    x.dateFrom,
+                    x.dateTo,
+                    x.time,
+                    x.price,
+                    new SubjectDto(x.subject.id, x.subject.name),
+                    new User(x.teacher?.email,
+                        x.teacher?.hashedPwd,
+                        x.teacher?.firstName,
+                        x.teacher?.lastName,
+                        x.teacher?.address,
+                        x.teacher?.phone),
+                ),
+            ),
+            count,
+            hasNext,
+            hasPrev);
+    }
+
+    async addAdvert(advert: AddAdvertDto): Promise<AdvertDto | undefined> {
+        const entity = AdvertEntity.create(new AdvertEntity(
+            advert.dateFrom,
+            advert.dateTo,
+            advert.place,
+            advert.level,
+            advert.price,
+            advert.time,
+            advert.teacherId,
+            advert.subjectId));
+
+        const created = await this.advertRepository.save(entity);
+        const subject = await this.subjectService.findById(created.subjectId);
+        const teacher = await this.userService.findById(created.teacherId);
+
+        return new AdvertDto(
+            created.id,
+            created.place,
+            advert.level,
+            advert.dateFrom,
+            advert.dateTo,
+            created.time,
+            created.price,
+            subject,
+            teacher);
+    }
+
+    async deleteAdvert(id: string, userId: string): Promise<boolean> {
+        const entity = await this.advertRepository.findOne(id);
+        if (entity != null) {
```

Comment:
`!==` bezpieczniej

Reply:
a bezpieczniej `if (entity)` nie jestem pewny ale jak nie znajdzie to dostaniejsz undefined a nie null